### ISSUE TITLE
We need to use garbageCleanup() for a buttload more objects / High GC Failure rate

### ISSUE THREAD

**Issue by RemieRichards** [sawu-tg/dragonroll#65]
Example Sdel controller stat outputs:
```Garbage | 0ds | (CPU:0%) | (Left: 128) | (Failed: 84.7981%)```
```Garbage | 0ds | (CPU:0%) | (Left: 7) | (Failed: 99.1755%)```

The game ends up with around 85-99% Failure (that is to say, forced Deletion) of each datum due to garbageCleanup() not being used to remove references (to be fair I only just implemented the changes to Sdel and added garbageCleanup() today 13/06/2015)

Should be a simple fix, just define proper working garbageCleanup()s for each type we have that has lots of references.

Amusingly, I have never seen it reach actual 100%, only 99% so SOMETHING out there is garbage collecting correctly!

**Comment by DaedalusGame** [comment]
explain how do garbage collect

do you get rid of the datum's references to other things or other thing's references to the datum?

**Comment by RemieRichards** [comment]
@DaedalusGame 
to get a datum to garbage collect it must not be REFERENCED (As in, another object --> this object) since the object is being garbage collected and is soon to not exist at all (and should be considered as no longer existing!) 

Obviously this isn't perfectly possible in some cases at the moment (something may reference src without src having a reference back to it, so in this case you need to GIVE src a reference to be able to clean up both later)

tl;dr whatever references point to the datum need to be nulled.

**Comment by DaedalusGame** [comment]
my kingdom for bidirectional references

**Comment by DaedalusGame** [comment]
you have saved the reference to the sdel'd object in a list, which kept the reference around and the object would never be garbage collected before it's force-deleted.

i did a poke at porting over the thing where it saves /ref[what] in the list instead like tg does it and i get 0% failures

butt butt butt butt butt butt butt

**Comment by RemieRichards** [comment]
@DaedalusGame I'm not sure how you'd remove it from the "deleted" list (also that was sawu not me :P) since you can't possibly know if it's going to GC within the 30 seconds of "patience" before it hard-dels, please pop up a PR and I'll review the fix.

**Comment by DaedalusGame** [comment]
MrPerson explained it in IRC
you keep track of an associative list with the string ref, instead of the actual ref, and the time when you gc'd so you can check if the object you're deleting is actually the object you sdel'd.

**Comment by RemieRichards** [comment]
Ah makes sense, so when it comes to harddel you have the ref to locate it again, instead of an ACTUAL reference upping it's reference count making it not GC.

**Comment by DaedalusGame** [comment]
yepp

**Comment by RemieRichards** [comment]
Heh, given that the 20-30 CPU we were getting when I fixed the main process loops, I wonder how low it'll be now that we're actually GCing? haha

**Comment by RemieRichards** [comment]
(Rename so I can close it when your fix is up bord :P)

**Comment by sawu-tg** [comment]
@RemieRichards  Gasp, who'da thunk we'd get a #BlameRemie

**Comment by RemieRichards** [comment]
@sawu-tg tis a dark day, also why assign yourself Bord's already fixed it apparently (although if it's not up in the next minute or two I'm doing it myself :P)

**Comment by DaedalusGame** [comment]
I had to do a few more tests dangit.

**Comment by sawu-tg** [comment]
@RemieRichards I didn't even mean to do that, nor did I notice. PAY NO ATTENTION.
